### ISSUE TITLE
Vault UI Feature Request: Don't Show all login methods in Vault UI Login Page

### ISSUE THREAD

**Issue by sgleske-ias** [hashicorp/vault#9076]
# Prior issues

This feature request was originally requested and partially implemented in #4307.

# Desired behavior

"Other" methods do not show up at all.  I would like to not expose unconfigured methods to users which causes confusion.

Only show auth methods which support showing via `unauth` visibility.

# Screenshot of current behavior

Here's a screenshot including the vault version.  "Other" tab is undesirable because it exposes all auth methods.

---

![Screenshot from 2020-05-25 14-44-16](https://user-images.githubusercontent.com/30084856/82837441-0d620780-9e97-11ea-92fa-9822f3d36898.png)

---

**Comment by ThomasKasene** [comment]
Good suggestion! I'd also like to enable admins to specify a default authentication method.

**Comment by slessardjr** [comment]
> Good suggestion! I'd also like to enable admins to specify a default authentication method.

Agreed, if we can't remove them I'd like to be able to make one the default.

**Comment by jmichelgarcia** [comment]
I would love to have this!

**Comment by erickufrin** [comment]
++1.

We need to be able to remove the 'Other' tab. It confuses our end users and causes issues.

**Comment by jppitout** [comment]
This would be great to alleviate confusion amongst our developers.

**Comment by mistial-dev** [comment]
If the issue is a "chicken and the egg problem", as mentioned [here](https://github.com/hashicorp/vault/issues/4307#issuecomment-379505408), then perhaps the best way to handle it would be a setting for "visible methods", rather than actually reading the auth mounts.

While that could mean that the UI and the backends get out of sync, that might actually be a desirable feature.  For example, we'd like to retain the capability to use root tokens for APIs if absolutely required, but would rather not show that as an option in the UI.

Hiding it would not technically stop it from being used, but it would serve to discourage the practice and reduce user confusion.

**Comment by itman-co** [comment]
I believe this is a must feature. Specially if you have a large number of people working with Vault.
Really confusing.

**Comment by davama** [comment]
Glad this FR exists.
Hope this gets looked at soon 

Thank you for the awesome application!

**Comment by rjhornsby** [comment]
@hellobontempo i'm a little bit confused by your reply. 

> we need to keep the dropdown so unauthenticated users are able to select any of the possible login methods.

We, the vault administrators, decide what login methods are available by configuring said authentication methods. It's unclear why all methods - even those explicitly and deliberately not configured and thus not available for use - must necessarily be shown in the list presented to end users.

Second, our vault 1.12 login screen doesn't look like your gif. There are no tabs and nothing called 'other'. What is being asked here is about a dropdown list.

![Screenshot 2023-04-05 at 12 50 40](https://user-images.githubusercontent.com/5077610/230163067-b46cd441-40bd-4fbe-86f6-c7f20f93771f.png)

![Screenshot 2023-04-05 at 12 50 46](https://user-images.githubusercontent.com/5077610/230162989-b88e13f9-8bd9-4cfc-b588-b65925932ba4.png)

**Comment by nneul** [comment]
@hellobontempo "We understand the login experience can be confusing, but need to keep the dropdown so unauthenticated users are able to select any of the possible login methods."    Why do unauthenticated users need to be able to choose an authentication method that we (as the admins of the server) do not WANT them to use, or even more applicably - have not even configured to be functional?

The option to list one explicitly is fine - but in your example - there are a whole slew of login methods that are not even configured to be functional. Why are they even shown?

![image](https://user-images.githubusercontent.com/828492/230163583-57c8384a-dbe9-41cc-966a-07204fff9491.png)

Example on my instance - I don't have GitHub or JWT or Radius or Username as configured/usable options. Why are you forcing my users to see those as choices when they can't even USE them?

**Comment by hellobontempo** [comment]
@rjhornsby to add methods as tabs, you have to select `List method when unauthenticated` when configuring the auth method. (I already had some of my methods enabled which is why they appeared in the gif) [docs here](https://developer.hashicorp.com/vault/api-docs/system/mounts#listing_visibility)

This configuration is then applied **to the method,** but is not user specific, which is why the dropdown is hardcoded to contain all possible login methods to Vault. Prior to authenticating, Vault has no knowledge of per-user capabilities

**Comment by patoarvizu** [comment]
@hellobontempo But Vault does have knowledge of what auth methods are available even when the user is unauthenticated, so couldn't it just trim the dropdown list to those options?

**Comment by slessardjr** [comment]
@hellobontempo maybe there is a misunderstanding, but for my users it was confusing to them why the default method wouldn't be OIDC.  It's extra clicks for them, the only users using any other auth methods are Admins.  

This feature request is to enhance end-user experience when logging in.  As an admin, I should be able to enforce which methods are the acceptable ones and which ones should be default.

**Comment by davama** [comment]
If you tune your prefered auth method as stated [here](https://developer.hashicorp.com/vault/docs/commands/auth/tune#listing-visibility)

ex:

`vault auth tune -listing-visibility="unauth" ldap/`

i believe the method gets selected by default upon an unauthenticated access to the UI

**Comment by slessardjr** [comment]
@davama that's exactly it. I have Ldap (on-prem ad linked), OIDC (Azure AD), and Token.  Ldap needs to be there for legacy reasons so I can't remove it, but trying to move users over to using Azure AD was failing because everyone was still just using LDAP.

**Comment by hellobontempo** [comment]
@slessardjr 
> Ldap needs to be there for legacy reasons so I can't remove it. 

Are you able to change the listing visibility for LDAP? This won't remove the mount, but it would remove it as a tab in the login form and point users to using the oidc azure option

**Comment by hellobontempo** [comment]
🪴 TL;DR 🪴 
Thank you for this discussion and we will be mindful of this feedback when planning future improvements surrounding the login flow. We understand it's an important feature to many of you, but at this time we are unable to just "filter the dropdown." I highly recommend leveraging the `listing_visibility` setting to expedite the user login flow (directions below). 

<hr> 
There are a couple different issues mentioned in this thread that are worth addressing separately:

1. user specific login views 👉  this is not possible as prior to authenticating there is no way for vault to know who is attempting to login 🕵️ 

2. customizing the dropdown 👉 
> @patoarvizu: Vault does have knowledge of what auth methods are available even when the user is unauthenticated, so couldn’t it just trim the dropdown list to those options?

The dropdown displays different auth method **types** not auth **mounts**. Vault does not actually have knowledge of which methods **types** are enabled, prior to authentication. The login form relies on [this](https://developer.hashicorp.com/vault/api-docs/system/internal-ui-mounts#get-available-visible-mounts) unauthenticated endpoint. 

```
$ curl \
    http://127.0.0.1:8200/v1/sys/internal/ui/mounts
```
Auth **mounts** (not auth method **types**) tuned to have `listing_visibility: unauth` are returned by the above request. This is how we add the tabs for an auth **mount** in the login form. Leveraging this param will ✨ save your user clicks ✨  if they select the tab of their preferred login mount and save the URL as a bookmark 😄  

<hr> 

#### UI Steps for changing the listing visibility of an auth **mount**:


1. select the mount and click `Configure`
<img width="400" alt="Screenshot 2023-04-05 at 3 38 47 PM" src="https://user-images.githubusercontent.com/68122737/230217869-443d89de-d2c5-4c3b-a9a6-241fe78539a7.png">

2. click `Method Options`   
<img width="400" alt="Screenshot 2023-04-05 at 3 37 35 PM" src="https://user-images.githubusercontent.com/68122737/230217780-834e6a0b-78f6-4e6b-be2c-8d377a4bc02f.png">
 3. check "List method when unauthenticated" and save! 💾 
 
<img width="400" alt="Screenshot 2023-04-05 at 3 40 28 PM" src="https://user-images.githubusercontent.com/68122737/230218246-84d694ae-150f-4e5b-aa41-93152ec58069.png">

**Comment by patoarvizu** [comment]
@hellobontempo Gotcha. So, with the understanding that you don't plan on working on this for now, let me ask the following question for my own education: I can see that the call to `/internal/ui/mounts` returns the list of auth **mounts**, but the information about each mount's **type** is also there, so that could be used to have logic in the front end to filter out the drop down list, wouldn't it? I'm not a front-end developer so I don't want to say how hard or easy it would be, but I know that the information about what auth types are available is there, it's unrelated to the authentication status of the user and that you're already doing some logic with that information to create the tabs.

I don't want to speak for everybody else here, but I don't think the request/wish is to have a dynamic list of auth types available to a yet-unauthorized user (I understand that's next to impossible), but rather just a list of available auth options overall, which doesn't need to be at the user level.

Anyway, thanks for taking the time to reply!

**Comment by rjhornsby** [comment]
>  a list of available auth options overall, which doesn't need to be at the user level

Correct. Like @nneul points out, methods like Github and RADIUS are not enabled at any level in our vault and the ask is to not have them shown at all.

**Comment by nneul** [comment]
I guess what I'm not getting on all of this - the "listing_visibility" setting already has a default of "hidden" and that's already controlling the display when user is unauthenticated. Why not just just "hide it" if that is the setting and there is any listing_visibility set to 'unauth'. Or if you want to have it be highly explicit, add an "invisible" value to choose to not have that auth type be shown at all, even in the other dropdown.

This cannot possibly be more than a few lines of code.

**Comment by hsimon-hashicorp** [comment]
While we love to see an energetic discussion, I do want to take an opportunity to remind folks of the [HashiCorp Community Guidelines](https://www.hashicorp.com/community-guidelines). Please remember that we're all human beings behind these screens, and that the Vault developers strive to create the best product they can for our users. Also, with a complex product like Vault, what might seem to be simple changes can have unintended consequences.
Above all else, please be kind and we shall do the same. Thank you! :)

**Comment by hellobontempo** [comment]
@patoarvizu - Thank you for your question! 😄  Essentially this feature request is more than a simple UI change, but a larger ask that requires lots of cross-team planning to thoughtfully implement.

I understand the main request here is "a dynamic list of auth types". Yes - `type` is a param returned by that `/sys/internal/ui/mounts` endpoint. But this information is still only returned when mounts are tuned to have `listing_visibility: unauth` and - as made apparent by this thread - this is a param many users are still unaware of 😅 

Therefore, filtering the dropdown using existing resources would filter the dropdown for **every** Vault user. The UI would be unusable for anyone who failed to tune their mount accordingly. (In addition, this would likely cause further confusion since it would _look_ like Vault has some awareness of available auth method types, but it would still only be aware of which mounts had been tuned from `hidden` to `unuath`)

This means to avoid any breaking changes dropdown customization is reliant on either:
1.) customizing the Vault UI which is unprecedented and would be part of a much larger initiative
2.) adding an additional config parameter to tune auth mounts - this is outside the scope of the ui team and would need to be planned in tandem with backend folks

Hopefully this explains why this isn't a straightforward feature request - though we do recognize there's a lot of support for it! 

Utilizing the tabs seems much easier than bothering with the dropdown, but I'm interested in hearing if/why this is not a viable solution. Any insight would help to prioritize this work 🧙

**Comment by pascal-hofmann** [comment]
@hellobontempo How can I configure a custom logo/text for the login button as seen in your gif above?

![Bildschirmfoto 2023-04-11 um 18 13 39](https://user-images.githubusercontent.com/2102878/231224664-b45849b3-1942-4f42-8f08-289f0ef728a6.png)

**Comment by finnstech** [comment]
@hellobontempo tracking internally with other UI/UX asks. Thanks for the feedback!

**Comment by Markus-** [comment]
I still don't get ist.

As I understood the most comments the only thing we want to have is: 

The admins, who configured the valid authentication methods want to have a config-parameter, where the shortcodes of these authentication methods are listed and if this config parameter is set the ui just show the methods in this list and does not check any other "backend"-stuff.

**Comment by heatherezell** [comment]
Hello folks! We appreciate the feedback here. Our product managers are looking for folks willing to chat in deeper detail (ie via Zoom). Please email me (my username here, at hashicorp dot com) if you're willing to set up a deeper-dive conversation around your needs.

**Comment by glerchundi** [comment]
Hi @heatherezell, I sent an email but got it back bounced. What I'm missing? 🤔 

<img width="466" alt="Screenshot 2024-11-13 at 22 33 10" src="https://github.com/user-attachments/assets/2aa7adb0-36de-4de9-91f0-42adc9a927f3">

**Comment by heatherezell** [comment]
> Hi @heatherezell, I sent an email but got it back bounced. What I'm missing? 🤔

Sometimes Google doesn't mind a missing dot, sometimes it does. Try it with a dot in between my first and last name?

**Comment by heatherezell** [comment]
> > Hello folks! We appreciate the feedback here. Our product managers are looking for folks willing to chat in deeper detail (ie via Zoom). Please email me (my username here, at hashicorp dot com) if you're willing to set up a deeper-dive conversation around your needs.
> 
> Is it possible to give you some ideas without scheduling a meeting?

Yes, we can absolutely do it async as well - there might be follow-up questions, though. :)

**Comment by lusu007** [comment]
> > > Hello folks! We appreciate the feedback here. Our product managers are looking for folks willing to chat in deeper detail (ie via Zoom). Please email me (my username here, at hashicorp dot com) if you're willing to set up a deeper-dive conversation around your needs.
> > 
> > 
> > Is it possible to give you some ideas without scheduling a meeting?
> 
> Yes, we can absolutely do it async as well - there might be follow-up questions, though. :)

Follow-up questions are welcome. I will send a few ideas to you in the next days. 😊

**Comment by Markus-** [comment]
Hi,

last year I posted a comment and I thought that I summed up the main issue quite good.

"The admins, who configured the valid authentication methods want to have a config-parameter, where the shortcodes of these authentication methods are listed and if this config parameter is set the ui just show the methods in this list and does not check any other "backend"-stuff."

We just want a possibility to disable the unconfigured authentication methods for the plain user.

Probably your devs can come up with an implementation idea which is not unrealistic and we can discuss this in the open.

Regards
Markus

**Comment by nneul** [comment]
I would add one additional one to @Markus- comment - I'd like to be able to additionally hide _configured_ auth endpoints from the displayed login panel UI. Even though they are enabled and configured, I don't necessarily want them presented to end users as a choice. Hiding _unconfigured_ endpoints should be the bare minimum. 

i.e. I'd like to show JUST the primary SAML/OIDC login, and not the ones for various other services like gitlab, ldap, etc. -- even if they are enabled.

**Comment by Markus-** [comment]
I implied this with the "shortcodes", so the idea is to have a parameter

visible_auth_method: ['oidc', 'ldap', 'token']

The listed methods must be configured (of course), but the list must not contain ALL configured methods.

**Comment by rjhornsby** [comment]
> i.e. I'd like to show JUST the primary SAML/OIDC login, and not the ones for various other services like gitlab, ldap, etc. -- even if they are enabled.

I think this is a good summary, and seems straight forward. Does the product management team have specific questions about this that we could answer?

**Comment by heatherezell** [comment]
Hello everyone! Thank you so much for being patient with us. Our product and design teams have been talking with users, and they have also put together a [quick Google survey](https://docs.google.com/forms/d/e/1FAIpQLScfMB0gMJ2SlOItU0GozGI373KwkmFyz1sqj8hE7bSwphE2VA/viewform) to gather your feedback. This has been a long time in the making, I know, but above all - we want to be in a position to do it right, which means any and all of your feedback is welcome. Thank you all in advance for your input!

**Comment by 7E6D4309-D2A9-4155-9D4F-287B8CDA14C1** [comment]
lol of course after all of this it ended up as an enterprise only feature.
